File: minmum_window_substring/MinimumWindowSubstring.java
public class MinimumWindowSubstring {
    public String minWindow(String s, String t) {
        if (s == null || s.length() == 0 || t == null || t.length() == 0) {
            return "";
        }
        
        Map<Character, int[]> charCounters = new HashMap<>();
        for (int i = 0; i < t.length(); i++) {
            char ch = t.charAt(i);
            if (!charCounters.containsKey(ch)) {
                charCounters.put(ch, new int[]{0});
            }
            charCounters.get(ch)[0]++;
        }
        
        int matches = 0;
        int slow = 0;
        int len = Integer.MAX_VALUE;
        int startIdx = -1;
        for (int fast = 0; fast < s.length(); fast++) {
            char ch = s.charAt(fast);
            if (!charCounters.containsKey(ch)) {
                continue;
            }
            if (--charCounters.get(ch)[0] == 0) {
                matches++;
            }
            
            while (matches == charCounters.size() && slow <= fast) {
                if (fast - slow + 1 < len) {
                    len = fast - slow + 1;
                    startIdx = slow;
                }
                if (!charCounters.containsKey(s.charAt(slow))) {
                    slow++;
                    continue;
                }
                if (++charCounters.get(s.charAt(slow++))[0] == 1) {
                    matches--;
                }
            }
        }
        
        return len == Integer.MAX_VALUE ? "" : s.substring(startIdx, startIdx + len);
    }
}

File: minmum_window_substring/Minimum_Window_Substring.cpp
class Minimum_Window_Substring {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty()) {
            return "";
        }
        
        unordered_map<char, int> table;
        for (int i = 0; i < t.size(); i++) {
            table[t[i]]++;
        }
        
        int start_idx = 0;
        int min_len = INT_MAX;
        int slow = 0;
        int matches = 0;
        for (int fast = 0; fast < s.size(); fast++) {
            if (table.find(s[fast]) == table.end() || (--table[s[fast]] == 0 && ++matches < table.size())) {
                continue;
            }
            
            while (matches == table.size()) {
                if (min_len > fast - slow + 1) {
                    min_len = fast - slow + 1;
                    start_idx = slow;
                }
                if (table.find(s[slow]) == table.end()) {
                    slow++;
                    continue;
                }
                if (++table[s[slow++]] > 0) {
                    matches--;
                }
            }
        }
        
        return min_len == INT_MAX ? "" : s.substr(start_idx, min_len);
    }
};